### PR TITLE
Add new command 'aesb' step back for ESIL

### PR DESCRIPTION
**Pull request by RKX1209**
Firstly user can save trace session by 'aets+' command in ESIL mode.
Then, 'aesb' restore session saved by 'aets+' and execute forward until desired point.
It's a same logic as it of 'dsb' in debug mode.

### CODE REVIEW

File: libr/anal/session.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+R_API void r_anal_esil_session_free(RAnalEsilSession *session) {
+	free (session->data);
+	free (session);
+}
+
+R_API void r_anal_esil_session_list(RAnalEsil *esil) {
+	if (!esil || !esil->sessions) {
+		return;
+	}
+	eprintf ("%d\n", r_list_length (esil->sessions));
```
Comment (radare) [inline libr/anal/session.c]: you may use esil->anal->cb_printf ("..."); this is hooked to r_cons_printf via rcore

File: libr/anal/session.c [hunk 2]
```diff
@@ -0,0 +1,68 @@
+/* radare - LGPL - Copyright 2017 - rkx1209 */
```
Comment (radare) [inline libr/anal/session.c]: newline between comment and includes

File: libr/anal/session.c (hunk 1 above)
Comment (radare) [inline libr/anal/session.c]: and list the sessions, not just show how many of thems there are

File: libr/anal/session.c [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	eprintf ("%d\n", r_list_length (esil->sessions));
+}
+
+R_API RAnalEsilSession *r_anal_esil_session_add(RAnalEsil *esil) {
+	RAnalEsilSession *session;
+	ut32 i;
+	if (!esil) {
+		return NULL;
+	}
+	session = R_NEW0 (RAnalEsilSession);
```
Comment (radare) [inline libr/anal/session.c]: define the variable right here, and save 1 line

File: libr/anal/session.c [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		return NULL;
+	}
+	if (!esil->stack_addr) {
+		R_FREE (session);
+		return NULL;
+	}
+
+	session->addr = esil->stack_addr;
+	session->size = esil->stack_size;
+	session->data = R_NEWS0 (ut8 *, session->size);
```
Comment (radare) [inline libr/anal/session.c]: check if allocation works or not

Comment (radare) [inline libr/anal/session.c]: check if stack_size is > 0 before allocating

File: libr/anal/session.c [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+R_API void r_anal_esil_session_set(RAnalEsil *esil, RAnalEsilSession *session) {
+	if (!esil || !session) {
+		return;
+	}
+	ut32 i;
+	RListIter *iter;
+	RRegArena *arena;
+	/* Restore registers */
+	for (i = 0; i < R_REG_TYPE_LAST; i++) {
+		iter = session->reg[i];
```
Comment (radare) [inline libr/anal/session.c]: define the RRegArena* arena right here

File: libr/anal/session.c [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		arena = iter->data;
+		if (esil->anal->reg->regset[i].arena->bytes) {
+			memcpy (esil->anal->reg->regset[i].arena->bytes, arena->bytes, arena->size);
```
Comment (radare) [inline libr/anal/session.c]: ensure destination arena fits with the src arena.

and i think there are functions in reg/arena that do that already

File: libr/core/cmd_anal.c [hunk 7]
```diff
@@ -2851,6 +2861,30 @@ R_API int r_core_esil_step(RCore *core, ut64 until_addr, const char *until_expr)
...
+	RAnalEsil *esil = core->anal->esil;
+	RListIter *tail;
+	RAnalEsilSession *before;
+	ut64 pc, end, prev = 0;
+	const char *name = r_reg_get_name (core->anal->reg, R_REG_NAME_PC);
+	if (!esil || !(tail = r_list_tail (esil->sessions))) {
+		return 0;
+	}
+
+	before = (RAnalEsilSession *) tail->data;
```
Comment (radare) [inline libr/core/cmd_anal.c]: define the variable here

File: libr/core/cmd_anal.c [hunk 8]
```diff
@@ -2851,6 +2861,30 @@ R_API int r_core_esil_step(RCore *core, ut64 until_addr, const char *until_expr)
...
+	end = r_reg_getv (core->anal->reg, name);
```
Comment (radare) [inline libr/core/cmd_anal.c]: ut64 here

File: libr/core/cmd_anal.c [hunk 9]
```diff
@@ -3676,6 +3722,19 @@ static void cmd_anal_esil(RCore *core, const char *input) {
 			r_anal_esil_free (esil);
 			break;
 		}
+		case 's': // "aets"
+			switch (input[2]) {
+				case 0:
```
Comment (radare) [inline libr/core/cmd_anal.c]: wrong indent, case should be at the same level as switch

**Comment by radare** [comment]
on windows:
```
libr_anal.a(esil.obj) : error LNK2019: unresolved external symbol r_anal_esil_session_free referenced in function r_anal_esil_new [C:\projects\radare2-shvdd\build\binr\rasm2\rasm2@exe.vcxproj]

```

**Comment by RKX1209** [comment]
OK. I've fixed. But, I don't know what happen on Windows...

**Comment by XVilka** [comment]
There are still crashes on some Travis tests

**Comment by radare** [comment]
Can you add tests in r2r? Thanks!

**Pull request merged**